pkg/reconciler/v1alpha1/taskrun/taskrun.go

```diff
@@ -187,3 +283,70 @@ func (c *Reconciler) updateStatus(taskrun *v1alpha1.TaskRun) (*v1alpha1.TaskRun,
 	}
 	return newtaskrun, nil
 }
+
+// finalState returns true if the taskrun has already finished
+func finalState(status v1alpha1.TaskRunConditionType) bool {
+	switch status {
+	case "", v1alpha1.PendingState, v1alpha1.TriggeredState:
+		return false
+	}
+	return true
+}
+
+func (c *Reconciler) getBuild(namespace, name string) (*buildv1alpha1.Build, error) {
+	return c.buildLister.Builds(namespace).Get(name)
+}
+
+func (c *Reconciler) createBuild(namespace string, b *buildv1alpha1.Build) (*buildv1alpha1.Build, error) {
+	return c.BuildClientSet.BuildV1alpha1().Builds(namespace).Create(b)
+}
+
+func (c *Reconciler) deleteBuild(namespace, name string) error {
+	return c.BuildClientSet.BuildV1alpha1().Builds(namespace).Delete(name, &metav1.DeleteOptions{})
+}
+
+// makeBuild creates a build from the task, using the task's buildspec.
+func makeBuild(t v1alpha1.Task) (*buildv1alpha1.Build, error) {
+	if t.Spec.BuildSpec == nil {
+		return nil, fmt.Errorf("nil BuildSpec")
+	}
+	return &buildv1alpha1.Build{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      t.Name,
+			Namespace: t.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(&t, groupVersionKind),
+			},
+		},
+		Spec: *t.Spec.BuildSpec,
+	}, nil
```

Comment:
you have to check if the BuildSpec has a template and apply that template.
Applying the template includes:
 * checking type of template
 * getting appropriate template object using lister
 * apply the template (use ApplyTemplate() from Build?)

Reply:
I have currently omitted this change as I believe the Build controller will do this for us:
https://github.com/knative/build/blob/e69c21ed74a05d83d68fc15bacbc458807be8288/pkg/controller/build/controller.go#L338

I could definitely be misunderstanding something though, let me know if you still believe we need to call ApplyTemplate()